// starrail.js/dist/models/relic/Relic.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const RelicData_1 = __importDefault(require("./RelicData"));
/**
 * @en Relic
 */
class Relic {
    /**
     * @param data
     * @param client
     */
    constructor(data, client) {
        this.client = client;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.relicData = new RelicData_1.default(json.getAsNumber("ID"), this.client);
        this.level = json.getAsNumberWithDefault(0, "Level");
        const mainAffixId = json.getAsNumber("MainAffixID");
        this.mainStat = this.relicData.mainStatGroup.mainStats.find(mainStat => mainStat.id === mainAffixId);
        this.subStats = json.get("RelicSubAffix").mapArray((_, subAffix) => {
            const subAffixId = subAffix.getAsNumber("SubAffixID");
            const subStatData = this.relicData.subStatGroup.subStats.find(s => s.id === subAffixId);
            const count = subAffix.getAsNumber("Cnt");
            const steps = subAffix.getAsNumberWithDefault(0, "Step");
            return {
                subStatData,
                count,
                steps,
                value: subStatData.baseValue * count + subStatData.stepValue * steps,
            };
        });
    }
}
exports.default = Relic;
