// starrail.js/dist/models/character/CharacterData.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = __importDefault(require("../../errors/AssetsNotFoundError"));
const TextAssets_1 = __importDefault(require("../assets/TextAssets"));
const CombatType_1 = __importDefault(require("../CombatType"));
const Path_1 = __importDefault(require("../Path"));
const Skill_1 = __importDefault(require("./skill/Skill"));
const ImageAssets_1 = __importDefault(require("../assets/ImageAssets"));
/**
 * @en CharacterData
 */
class CharacterData {
    /**
     * @param id
     * @param client
     */
    constructor(id, client) {
        this.id = id;
        this.client = client;
        const _data = client.cachedAssetsManager.getStarRailCacheData("AvatarConfig")[this.id];
        if (!_data)
            throw new AssetsNotFoundError_1.default("Character", this.id);
        this._data = _data;
        const _itemData = client.cachedAssetsManager.getStarRailCacheData("ItemConfigAvatar")[this.id];
        if (!_itemData)
            throw new AssetsNotFoundError_1.default("Character Item", this.id);
        this._itemData = _itemData;
        const json = new config_file_js_1.JsonReader(this._data);
        const itemJson = new config_file_js_1.JsonReader(this._itemData);
        this.name = new TextAssets_1.default(json.getAsNumber("AvatarName", "Hash"), this.client);
        this.description = new TextAssets_1.default(itemJson.getAsNumber("ItemBGDesc", "Hash"), this.client);
        this.stars = Number(json.getAsString("Rarity").slice(-1));
        this.combatType = new CombatType_1.default(json.getAsString("DamageType"), this.client);
        this.path = new Path_1.default(json.getAsString("AvatarBaseType"), this.client);
        this.skills = json.get("SkillList").mapArray((_, skillId) => new Skill_1.default(skillId.getAsNumber(), this.client));
        this.icon = new ImageAssets_1.default(json.getAsString("DefaultAvatarHeadIconPath"), this.client);
        this.sideIcon = new ImageAssets_1.default(json.getAsString("AvatarSideIconPath"), this.client);
        this.miniIcon = new ImageAssets_1.default(json.getAsString("AvatarMiniIconPath"), this.client);
        this.teamActionIcon = new ImageAssets_1.default(json.getAsString("ActionAvatarHeadIconPath"), this.client);
        this.teamWaitingIcon = new ImageAssets_1.default(json.getAsString("WaitingAvatarHeadIconPath"), this.client);
        this.splashImage = new ImageAssets_1.default(json.getAsString("AvatarCutinFrontImgPath"), this.client);
        this.splashCutInFigureImage = new ImageAssets_1.default(json.getAsString("AvatarCutinImgPath"), this.client);
        this.splashCutInBackgroundImage = new ImageAssets_1.default(json.getAsString("AvatarCutinBgImgPath"), this.client);
        this.shopItemIcon = new ImageAssets_1.default(itemJson.getAsString("ItemAvatarIconPath"), this.client);
    }
}
exports.default = CharacterData;
