// starrail.js/dist/models/character/Character.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const CharacterData_1 = __importDefault(require("./CharacterData"));
const LightCone_1 = __importDefault(require("../light_cone/LightCone"));
const Relic_1 = __importDefault(require("../relic/Relic"));
const SkillTreeNode_1 = __importDefault(require("./skill/SkillTreeNode"));
/**
 * @en Character
 */
class Character {
    /**
     * @param data
     * @param client
     */
    constructor(data, client) {
        this.client = client;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.characterData = new CharacterData_1.default(json.getAsNumber("AvatarID"), this.client);
        this.lightCone = json.has("EquipmentID", "ID") ? new LightCone_1.default(json.getAsJsonObject("EquipmentID"), this.client) : null;
        this.relics = json.getAsJsonArrayWithDefault([], "RelicList").map(relic => new Relic_1.default(relic, this.client));
        this.level = json.getAsNumber("Level");
        this.exp = json.getAsNumberWithDefault(0, "EXP");
        this.ascension = json.getAsNumberWithDefault(0, "Promotion");
        this.eidolons = json.getAsNumberWithDefault(0, "Rank");
        this.skills = json.get("BehaviorList").mapArray((_, skill) => new SkillTreeNode_1.default(skill.getAsNumber("BehaviorID"), this.client).getSkillTreeNodeByLevel(skill.getAsNumber("Level")));
    }
}
exports.default = Character;
