// starrail.js/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchJSON = exports.User = exports.StatProperty = exports.Path = exports.CombatType = exports.RelicSubStatGroup = exports.RelicSetBonus = exports.RelicSet = exports.RelicMainStatGroup = exports.RelicExpType = exports.RelicData = exports.Relic = exports.LightConeSuperimposition = exports.LightConeExpType = exports.LightConeData = exports.LightCone = exports.CharacterData = exports.Character = exports.Skill = exports.LeveledSkill = exports.TextAssets = exports.ImageAssets = exports.RequestError = exports.AssetsNotFoundError = exports.StarRail = exports.ObjectKeysManager = exports.CachedAssetsManager = void 0;
const CachedAssetsManager_1 = __importDefault(require("./client/CachedAssetsManager"));
exports.CachedAssetsManager = CachedAssetsManager_1.default;
const ObjectKeysManager_1 = __importDefault(require("./client/ObjectKeysManager"));
exports.ObjectKeysManager = ObjectKeysManager_1.default;
const StarRail_1 = __importDefault(require("./client/StarRail"));
exports.StarRail = StarRail_1.default;
const AssetsNotFoundError_1 = __importDefault(require("./errors/AssetsNotFoundError"));
exports.AssetsNotFoundError = AssetsNotFoundError_1.default;
const RequestError_1 = __importDefault(require("./errors/RequestError"));
exports.RequestError = RequestError_1.default;
const ImageAssets_1 = __importDefault(require("./models/assets/ImageAssets"));
exports.ImageAssets = ImageAssets_1.default;
const TextAssets_1 = __importDefault(require("./models/assets/TextAssets"));
exports.TextAssets = TextAssets_1.default;
const LeveledSkill_1 = __importDefault(require("./models/character/skill/LeveledSkill"));
exports.LeveledSkill = LeveledSkill_1.default;
const Skill_1 = __importDefault(require("./models/character/skill/Skill"));
exports.Skill = Skill_1.default;
const Character_1 = __importDefault(require("./models/character/Character"));
exports.Character = Character_1.default;
const CharacterData_1 = __importDefault(require("./models/character/CharacterData"));
exports.CharacterData = CharacterData_1.default;
const LightCone_1 = __importDefault(require("./models/light_cone/LightCone"));
exports.LightCone = LightCone_1.default;
const LightConeData_1 = __importDefault(require("./models/light_cone/LightConeData"));
exports.LightConeData = LightConeData_1.default;
const LightConeExpType_1 = __importDefault(require("./models/light_cone/LightConeExpType"));
exports.LightConeExpType = LightConeExpType_1.default;
const LightConeSuperimposition_1 = __importDefault(require("./models/light_cone/LightConeSuperimposition"));
exports.LightConeSuperimposition = LightConeSuperimposition_1.default;
const Relic_1 = __importDefault(require("./models/relic/Relic"));
exports.Relic = Relic_1.default;
const RelicData_1 = __importDefault(require("./models/relic/RelicData"));
exports.RelicData = RelicData_1.default;
const RelicExpType_1 = __importDefault(require("./models/relic/RelicExpType"));
exports.RelicExpType = RelicExpType_1.default;
const RelicMainStatGroup_1 = __importDefault(require("./models/relic/RelicMainStatGroup"));
exports.RelicMainStatGroup = RelicMainStatGroup_1.default;
const RelicSet_1 = __importDefault(require("./models/relic/RelicSet"));
exports.RelicSet = RelicSet_1.default;
const RelicSetBonus_1 = __importDefault(require("./models/relic/RelicSetBonus"));
exports.RelicSetBonus = RelicSetBonus_1.default;
const RelicSubStatGroup_1 = __importDefault(require("./models/relic/RelicSubStatGroup"));
exports.RelicSubStatGroup = RelicSubStatGroup_1.default;
const CombatType_1 = __importDefault(require("./models/CombatType"));
exports.CombatType = CombatType_1.default;
const Path_1 = __importDefault(require("./models/Path"));
exports.Path = Path_1.default;
const StatProperty_1 = __importDefault(require("./models/StatProperty"));
exports.StatProperty = StatProperty_1.default;
const User_1 = __importDefault(require("./models/User"));
exports.User = User_1.default;
// functions
var axios_utils_1 = require("./utils/axios_utils");
Object.defineProperty(exports, "fetchJSON", { enumerable: true, get: function () { return axios_utils_1.fetchJSON; } });
